kernel/src/sched.rs

```diff
@@ -373,28 +473,62 @@ impl Kernel {
         }
```

Comment:
With this implementation, the scheduler cannot force the chip to go to sleep. I think at least the `&& self.processes_blocked()` check needs to be removed, otherwise a power scheduler (say) could not decide to sleep while a process is runnable if it is trying to save energy.

For interrupts (and software interrupts), I don't think the decision is as clear. If a scheduler ignores pending interrupts, will the chip just immediately wake up (or not sleep at all) on some (all?) platforms? Also, if there are no other pending interrupts, a scheduler could cause the system to stop as it would never wake up again.

I think making a conscious decision to say that the scheduler (at least right now) cannot ignore pending interrupts to go to sleep is reasonable, but we should document that.

Reply:
I have pushed an update that resolves this by removing the nested loop so that there is only a single main loop that matches whether to perform kernel work or try to run a process / sleep. I actually think that a design like this might be cleaner than what I just pushed:
```
/// Enum representing the actions the scheduler can request in each call to scheduler.next()
#[derive(Copy, Clone)]
pub(crate) enum SchedulingDecision {
    /// Tell the kernel loop to call execute_kernel_tasks()
    KernelWork,

    /// Tell the kernel to run the specified process with the passed timeslice.
    /// If `None` is passed as a timeslice, the process will be run cooperatively.
    RunProcess((AppId, Option<u32>)),

    /// Tell the kernel to go to sleep. Notably, if the scheduler asks the kernel
    /// to sleep when kernel tasks are ready, the kernel will not sleep, and will
    /// instead restart the main loop and call `next()` again.
    Sleep,
}

impl Kernel {
    unsafe fn kernel_work_ready(&self, chip: &C) -> bool {
        chip.has_pending_interrupts()
            || DynamicDeferredCall::global_instance_calls_pending().unwrap_or(false)
    }

    fn kernel_loop(...) {
        loop {
            chip.watchdog().tickle();
            unsafe {
                scheduler.execute_kernel_work(chip);
                match scheduler.next(self, kernel, chip) {
                    KernelWork => scheduler.execute_kernel_work(chip),
                    RunProcess((appid, timeslice)) => {
                        self.process_map_or((), appid, |process| {
                            let (reason, time_executed) = self.do_process(
                                platform,
                                chip,
                                scheduler,
                                process,
                                ipc,
                                timeslice_us,
                            );
                            scheduler.result(reason, time_executed);
                        });
                    }
                    Sleep => {
                        chip.atomic(|| {
                            // cannot sleep if interrupts are pending, as on most platforms
                            // unhandled interrupts will wake the device.
                            if !chip.has_pending_interrupts()
                                && !DynamicDeferredCall::global_instance_calls_pending()
                                    .unwrap_or(false)
                            {
                                chip.watchdog().suspend();
                                chip.sleep();
                                chip.watchdog().resume();
                            }
                        });
                    }
                }
            }
        }
    }
}
```

But then we still rely on each scheduler implementation calling `kernel.kernel_work_read()` within `next()`, rather than the cleaner trait function with a default implementation.  Thoughts?

Reply:
What's the purpose of the `scheduler.execute_kernel_work(chip);` _before_ the call to `next`? Couldn't that be the call to `kernel_work_ready`?

Reply:
It shouldn't be there, I just forgot to delete that line. Edited!

Reply:
I think I prefer the current approach in the PR, but with next() returning an enum like you have in the comment (but with only run and sleep as options). That way the structure of doing kernel work first is preserved, and a simpler scheduler doesn't have to worry about getting that right.

It also keeps the narrow scope of `next()`: the kernel wants to run a process, the scheduler has to say which one or none. What happens if a scheduler _really_ wants to prioritize kernel work? Well, I'm not sure Tock is a great fit for that. At some point, we can't just continuously check if an interrupt has triggered. Once a scheduler returns `RunProcess`, the kernel will call `continue_process()` before actually running anything, so the scheduler could stop there and resume kernel work. If an interrupt fires after then but before the actual context switch, the interrupt won't be handled until after the process runs. I think that is reasonable.

Reply:
Yeah, and the scheduler can redefine `execute_kernel_work()` to loop if required, so I think this design is fine. I pushed an update that uses an enum with only run and sleep options.